staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -92,10 +92,12 @@ func StreamObject(statusCode int, gv schema.GroupVersion, s runtime.NegotiatedSe
 // The context is optional and can be nil.
 func SerializeObject(mediaType string, encoder runtime.Encoder, w http.ResponseWriter, req *http.Request, statusCode int, object runtime.Object) {
 	w.Header().Set("Content-Type", mediaType)
-	w.WriteHeader(statusCode)
 
-	if err := encoder.Encode(object, w); err != nil {
-		errorJSONFatal(err, encoder, w)
+	if output, err := runtime.Encode(encoder, object); err != nil {
```

Comment:
This means we will no longer serialize large responses (like `list 500 nodes`, `list 500 pods`, `list all secrets`, etc) in a streaming fashion.

@kubernetes/sig-scalability-pr-reviews @kubernetes/sig-api-machinery-pr-reviews might have opinions around losing that in order to get a better status code for something that is arguably a coding error.

Reply:
I think it would make more sense to wrap the responsewriter we pass to `encoder.Encode` to call `WriteHeader(statusCode)` on first `Write()` (basically the same thing response writer already does, but with our custom status code instead of an implicit 200). Then, as long as an encoder error doesn't output anything before failure (which it shouldn't if the object can't be encoded at all), we still have the ability to write a serialization error code/response in failure cases.

Reply:
Thanks, understood on the first point. Definitely don't think the tradeoff is worth it in that case. Good idea on the second point. I will look into getting that off the ground.

Reply:
If we're going to make this change, I'd suggest we run some scalability tests first to ensure api call latencies aren't affected much.

Reply:
@shyamjvs @liggitt I made a new commit that preserves the streaming fashion of serialization, while still allowing for more control over the header content. I believe this should take care of any scalability concern, but please let me know if there's tests I should run (I'm assuming e2e tests) to ensure this. Thanks very much @liggitt for the review and idea.